File: 3831-find-x-value-of-array-i/find-x-value-of-array-i.py
class Solution:
    def resultArray(self, nums: List[int], k: int) -> List[int]:
        result = [0] * k
        count = [0] * k
        
        for num in nums:
            next_count = [0] * k
            val = num % k
            next_count[val] += 1
            
            for r in range(k):
                if count[r] > 0:
                    next_count[(r * val) % k] += count[r]
                    
            for r in range(k):
                result[r] += next_count[r]
                count[r] = next_count[r]
                
        return result

File: 3831-find-x-value-of-array-i/find-x-value-of-array-i.java
class Solution {
    public long[] resultArray(int[] nums, int k) {
        long[] result = new long[k];
        long[] count = new long[k];
        
        for (int num : nums) {
            long[] nextCount = new long[k];
            int val = num % k;
            nextCount[val]++;
            
            for (int r = 0; r < k; r++) {
                if (count[r] > 0) {
                    nextCount[(r * val) % k] += count[r];
                }
            }
            
            for (int r = 0; r < k; r++) {
                result[r] += nextCount[r];
                count[r] = nextCount[r];
            }
        }
        
        return result;
    }
}